### PR TITLE
Feature/retrievewaterconsumptionobserveds

### PR DESCRIPTION
**Pull request by shestyf**

### CODE REVIEW

File: internal/pkg/_presentation/api/ngsi-ld/context/context.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	"github.com/rs/zerolog"
+	log "github.com/rs/zerolog/log"
+)
+
+type contextSource struct {
+	app application.Application
+	log zerolog.Logger
+}
+
+//CreateSource instantiates and returns a Fiware ContextSource that wraps the provided db interface
```
Comment (istyf) [inline internal/pkg/_presentation/api/ngsi-ld/context/context.go]: No db interface is being wrapped here anymore

File: internal/pkg/_presentation/api/ngsi-ld/context/context.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	if err != nil {
+		return err
+	}
+
+	observedAt, err := time.Parse(time.RFC3339, wco.WaterConsumption.ObservedAt)
+	if err != nil {
+		return err
+	}
+
+	err = cs.app.UpdateWaterConsumption(wco.ID, wco.WaterConsumption.Value, observedAt)
```
Comment (istyf) [inline internal/pkg/_presentation/api/ngsi-ld/context/context.go]: Use entityID here instead

File: internal/pkg/_presentation/api/ngsi-ld/context/context.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+func CreateSource(app application.Application, log zerolog.Logger) ngsi.ContextSource {
+	return &contextSource{
+		app: app,
+		log: log,
+	}
+}
+
+func (cs contextSource) CreateEntity(typeName, entityID string, req ngsi.Request) error {
+	sublogger := log.Logger
+	sublogger.With().Str("entityID", entityID).Str("entityType", typeName).Logger()
```
Comment (istyf) [inline internal/pkg/_presentation/api/ngsi-ld/context/context.go]: no need to create a sub logger here,  but if you want it you should derive it from the inserted logger and not from scratch. sublogger := cs.log.With()......Logger()

File: internal/pkg/_presentation/api/ngsi-ld/context/context.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+	observedAt, err := time.Parse(time.RFC3339, wco.WaterConsumption.ObservedAt)
+	if err != nil {
+		return err
+	}
+
+	err = cs.app.UpdateWaterConsumption(wco.ID, wco.WaterConsumption.Value, observedAt)
+	if err != nil {
+		return err
+	}
```
Comment (istyf) [inline internal/pkg/_presentation/api/ngsi-ld/context/context.go]: Remove this if-block.

File: internal/pkg/_presentation/api/ngsi-ld/context/context.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+	waterconsumptions, err := getWaterConsumptions(cs.app, query)
+	if err != nil {
+		return err
+	}
+
+	for _, w := range waterconsumptions {
+		entity := fiware.NewWaterConsumptionObserved(w.Device)
+
+		entity.WithConsumption(w.Device, w.Consumption, w.Timestamp)
```
Comment (istyf) [inline internal/pkg/_presentation/api/ngsi-ld/context/context.go]: Move WithConsumption to the same line as NewWCO above.

File: internal/pkg/_presentation/api/ngsi-ld/context/context_test.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/diwise/api-smartwater/internal/pkg/application"
+	"github.com/diwise/api-smartwater/internal/pkg/infrastructure/repositories/models"
+	"github.com/diwise/ngsi-ld-golang/pkg/ngsi-ld"
+	"github.com/matryer/is"
+	"github.com/rs/zerolog/log"
+)
+
+func TestUpdateWaterConsumption(t *testing.T) {
+
+	req, _ := http.NewRequest("POST", "http://localhost:8090/ngsi-ld/v1/entities", bytes.NewBuffer([]byte(wcoJson)))
```
Comment (istyf) [inline internal/pkg/_presentation/api/ngsi-ld/context/context_test.go]: Remove the protocol, host and port here

File: internal/pkg/infrastructure/repositories/database/database.go [hunk 7]
```diff
@@ -103,6 +104,29 @@ func NewDatabaseConnection(connect ConnectorFunc) (Datastore, error) {
...
+func (db *myDB) GetWaterConsumptions(deviceId string, from, to time.Time, limit uint64) ([]models.WaterConsumption, error) {
+	wcos := []models.WaterConsumption{}
+	gorm := db.impl.Order("timestamp")
+
+	if deviceId != "" {
+		gorm = gorm.Where("device = ?", deviceId)
+	}
+
+	if !from.IsZero() || !to.IsZero() {
+		gorm = insertTemporalSQL(gorm, "timestamp", from, to) //make sure to set to Descending order
```
Comment (istyf) [inline internal/pkg/infrastructure/repositories/database/database.go]: Make sure to set descending order where and when? I think we need a test that shows the behaviour of this function when from and to are zero, but there is a limit. That might be when you need descending order in the query.

**Review by istyf — CHANGES_REQUESTED** [review]
needs some more work

File: internal/pkg/_presentation/api/ngsi-ld/context/context.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+	if err != nil {
+		return err
+	}
+
+	observedAt, err := time.Parse(time.RFC3339, wco.WaterConsumption.ObservedAt)
+	if err != nil {
+		return err
+	}
+
+	err = cs.app.UpdateWaterConsumption(entityID, wco.WaterConsumption.Value, observedAt)
```
Comment (istyf) [inline internal/pkg/_presentation/api/ngsi-ld/context/context.go:47]: You could return immediately here. No point in saving the return value in a variable and then return that on the next line.

**Review by istyf — APPROVED** [review]
/lgtm
